homeassistant/components/tibber/sensor.py

```diff
@@ -328,14 +403,93 @@ async def async_setup_entry(
     async_add_entities(entities, True)
 
 
-class TibberSensor(SensorEntity):
-    """Representation of a generic Tibber sensor."""
+async def _async_setup_data_api_sensors(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up sensors backed by the Tibber Data API."""
+
+    runtime = hass.data.get(DOMAIN)
+    if runtime is None:
+        raise PlatformNotReady("Tibber runtime is not ready")
+    coordinator = TibberDataAPICoordinator(hass, entry, runtime)
+
+    await coordinator.async_config_entry_first_refresh()
+
+    entities: list[TibberDataAPISensor] = []
+    api_sensors = {sensor.key: sensor for sensor in DATA_API_SENSORS}
+
+    for device in coordinator.data.values():
+        for sensor in device.sensors:
+            description: SensorEntityDescription | None = api_sensors.get(sensor.id)
+            if description is None:
+                _LOGGER.debug(
+                    "Sensor %s not found in DATA_API_SENSORS, skipping", sensor
+                )
+                continue
+            entities.append(
+                TibberDataAPISensor(
+                    coordinator, device, description, sensor.description
+                )
+            )
+    async_add_entities(entities)
+
+
+class TibberDataAPISensor(CoordinatorEntity[TibberDataAPICoordinator], SensorEntity):
+    """Representation of a Tibber Data API capability sensor."""
 
     _attr_has_entity_name = True
 
     def __init__(
-        self, *args: Any, tibber_home: tibber.TibberHome, **kwargs: Any
+        self,
+        coordinator: TibberDataAPICoordinator,
+        device: TibberDevice,
+        entity_description: SensorEntityDescription,
+        name: str,
     ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+
+        self._device_id: str = device.id
+        self.entity_description = entity_description
+        self._attr_name = name
+
+        self._attr_unique_id = f"{device.external_id}_{self.entity_description.key}"
+
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, device.external_id)},
+            name=device.name,
+            manufacturer=device.brand,
+            model=device.model,
+        )
+
+    @property
+    def native_value(
+        self,
+    ) -> StateType:
+        """Return the value reported by the device."""
+        device = self.coordinator.data.get(self._device_id)
+        if device is None:
+            return None
+
+        for sensor in device.sensors:
+            if sensor.id == self.entity_description.key:
```

Comment:
It would be nicer if the coordinator from this list can make a dict that maps the sensor id to the sensor instead of forcing all entities to iterate the list when updating state. Then each sensor can simply look up its value in the dict using its description key.